Rubberduck.VBEEditor/ComManagement/TypeLibs/TypeLibs.cs

```diff
@@ -0,0 +1,1669 @@
+        {
+            _wrappedObjectRCW = rawTypeInfo;
+            InitCommon();
+        }
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="rawObjectPtr">The raw unmanaged pointer to the ITypeInfo</param>
+        /// <param name="parentUserFormUniqueId">used internally for providing a name for UserForm base classes</param>
+        public TypeInfoWrapper(IntPtr rawObjectPtr, int? parentUserFormUniqueId = null)
+        {
+            _rawObjectPtr = rawObjectPtr;
+            
+            // We have to restrict interface requests to VBE hosted ITypeInfos due to a bug in their implementation.
+            // See TypeInfoWrapper class XML doc for details.
+            _ITypeInfo_Aggregator       = new RestrictComInterfaceByAggregation<ComTypes.ITypeInfo>(rawObjectPtr, queryForType: false);
+            _IVBEComponent_Aggregator   = new RestrictComInterfaceByAggregation<IVBEComponent>(rawObjectPtr);
+            _IVBETypeInfo_Aggregator    = new RestrictComInterfaceByAggregation<IVBETypeInfo>(rawObjectPtr);
+
+            // base classes of VBE UserForms cause an access violation on calling GetDocumentation(MEMBERID_NIL)
+            // so we have to detect UserForm parents, and ensure GetDocumentation(MEMBERID_NIL) never gets through
+            if (parentUserFormUniqueId.HasValue)
+            {
+                _cachedTextFields = new TypeLibTextFields { _name = "_UserFormBase{unnamed}#" + parentUserFormUniqueId };
+            }
+
+            InitCommon();
+            DetectUserFormClass();
+        }
+
+        private bool _isDisposed;
+        public void Dispose()
+        {
+            if (_isDisposed) return;
+            _isDisposed = true;
+
+            _typeInfosWrapped?.Dispose();
+            _containerTypeLib?.Dispose();
+            _ITypeInfo_Aggregator?.Dispose();
+            _IVBEComponent_Aggregator?.Dispose();
+            _IVBETypeInfo_Aggregator?.Dispose();
+        }
+
+        private TypeLibTextFields? _cachedTextFields;
+        TypeLibTextFields CachedTextFields
+        {
+            get
+            {
+                if (!_cachedTextFields.HasValue)
+                {
+                    var cache = new TypeLibTextFields();
+                    target_ITypeInfo.GetDocumentation((int)TypeLibConsts.MEMBERID_NIL, out cache._name, out cache._docString, out cache._helpContext, out cache._helpFile);
+                    _cachedTextFields = cache;
+                }
+                return _cachedTextFields.Value;
+            }
+        }
+
+        public string Name { get => CachedTextFields._name; }
+        public string DocString { get => CachedTextFields._docString; }
+        public int HelpContext { get => CachedTextFields._helpContext; }
+        public string HelpFile { get => CachedTextFields._helpFile; }
+
+        public string GetProgID() => (Container?.Name ?? "") + "." + CachedTextFields._name;
```

Comment:
Seems to me that it would create names like `.MyUserForm`? Probably should only add `.` if there's a container name?

Reply:
i think it would be clearer to add {unnamedlibrary}.